fastfield_codecs/src/linearinterpol.rs

```diff
@@ -0,0 +1,279 @@
+                offset = offset.max(calculated_value - actual_value);
+            } else {
+                //positive value no offset reuqired
+                rel_positive_max = rel_positive_max.max(actual_value - calculated_value);
+            }
+        }
+
+        // rel_positive_max will be adjusted by offset
+        let relative_max_value = rel_positive_max + offset;
+
+        let num_bits = compute_num_bits(relative_max_value);
+        let mut bit_packer = BitPacker::new();
+        for (pos, val) in data_iter.enumerate() {
+            let calculated_value = get_calculated_value(first_val, pos as u64, slope);
+            let diff = (val + offset) - calculated_value;
+            bit_packer.write(diff, num_bits, write)?;
+        }
+        bit_packer.close(write)?;
+
+        let footer = LinearInterpolFooter {
+            relative_max_value,
+            offset,
+            first_val,
+            last_val,
+            num_vals: stats.num_vals,
+            min_value: stats.min_value,
+            max_value: stats.max_value,
+        };
+        footer.serialize(write)?;
+        Ok(())
+    }
+}
+fn get_slope(first_val: u64, last_val: u64, num_vals: u64) -> f64 {
+    (last_val as f64 - first_val as f64) / (num_vals as u64 - 1) as f64
+}
+
+fn get_calculated_value(first_val: u64, pos: u64, slope: f64) -> u64 {
+    first_val + (pos as f64 * slope) as u64
+}
+impl FastFieldSerializerEstimate for LinearInterpolFastFieldSerializer {
+    /// estimation for linear interpolation is hard because, you don't know
+    /// where the local maxima are for the deviation of the calculated value and
+    /// the offset is also unknown.
+    fn estimate(fastfield_accessor: &impl FastFieldDataAccess, stats: FastFieldStats) -> f32 {
+        let first_val = fastfield_accessor.get(0);
+        let last_val = fastfield_accessor.get(stats.num_vals as u32 - 1);
+        let slope = get_slope(first_val, last_val, stats.num_vals);
+
+        // let's sample at 10%, 20%, 25%, 50%, 75%, 90%
+        let num_vals = stats.num_vals as f32 / 100.0;
+        let sample_positions = [
+            (num_vals * 10.0) as usize,
+            (num_vals * 20.0) as usize,
+            (num_vals * 25.0) as usize,
+            (num_vals * 50.0) as usize,
+            (num_vals * 75.0) as usize,
+            (num_vals * 90.0) as usize,
+        ];
+
+        let max_distance = sample_positions
+            .iter()
+            .map(|pos| {
+                let calculated_value = get_calculated_value(first_val, *pos as u64, slope);
+                let actual_value = fastfield_accessor.get(*pos as u32);
+                distance(calculated_value, actual_value)
+            })
+            .max()
+            .unwrap();
+
+        // the theory would be that we don't have the actual max_distance, but we are close within 50%
+        // threshold.
+        // It is multiplied by 2 because in a log case scenario the line would be as much above as
```

Comment:
I agree, but log isn't the bad scenario is it?

Reply:
It's also not the best case. I think the estimation is reasonable for a lot of cases, but it could make sense to add a worst case estimation?